### DISCUSSION TITLE
ESPHome / platformIO project

### DISCUSSION THREAD

**Discussion by fabianmuehlberger** [Seeed-Solution/SenseCAP_Indicator_ESP32#35] [Ideas]
Hey, 

Are there any plans for creating an ESPHome project for the Indicator?

I am currently exploring the capabilities of the SenseCap Indicator and was curious if a ESPHome implementation would be a: Possible, and b: A goal worth pursuing.

I'm already a bit familiar with the ESPHome ecosystem, and since the Indicator is using the esp-idf I was wondering what it would take to make it an esphome device.
The esphome integration seems to be a perfect match, making the integration easier, since the source can be compiled within home assistant, as well as providing a robust communication between HA and the device via a native API. The ability to create custom buttons and actions with just one configuration file would be wonderful.

I assume that the main hurdle would be to implement a custom display component for esphome, but there is already some work done:  [esphome-gui by lukasz-tuz](https://github.com/lukasz-tuz/esphome-gui).

Best 
Fabian

**Comment by Love4yzp** [comment]
Hi Fabian,

Thanks for your interesting query about integrating the SenseCAP Indicator with ESPHome. It's always exciting to see innovative ideas for our products.

From my understanding and maintenance experience, integrating the SenseCAP indicator with ESPHome could be a bit complex. The indicator has two chips, ESP32S3 and RP2040. As per the Diagram below, the ESP32 is used for screen display and wireless communications (WiFi, Bluetooth, LoRa), while the RP2040 handles the SDCard and connects external devices through GPIO (UART, IIC, Analog). The ESP32 and RP2040 communicate via UART.

In existing projects I've worked on, the RP2040 often connects to multiple IIC devices via an IIC expansion board. One of the Grove GPIO ports can be used in ADC mode, or alternatively, GPIO can be switched to UART mode for communication with other MCUs. The RP2040 then sends data to the ESP32 for further processing, including network connectivity, data requests, and screen display interactions (typically using LVGL as it's already ported: [Low Code UI Design for SenseCAP Indicator with SquareLine](https://www.hackster.io/spenyan/low-code-ui-design-for-sensecap-indicator-with-squareline-9825fe)).

As for integrating with ESPHome (which I'm not overly familiar with but will try out, as I happen to have a XIAOS3 – haha), I'm curious about how you would handle the screen driver in this setup. **Configuring WiFi and Bluetooth might be straightforward, but communicating with the RP2040 via UART and managing the screen display could pose challenges. How do you envision tackling these aspects in ESPHome?**

I'm quite intrigued by your approach, especially the use of pre-compiled firmware and configuration files. I'll definitely spend some time learning about this, as I have seen similar projects like [SenseCapConfig on GitHub](https://github.com/YooSHome/SenseCapConfig), where replacing configuration files and working directly with the server seems to be an interesting approach.

![480](https://files.seeedstudio.com/wiki/SenseCAP/SenseCAP_Indicator/SenseCAP_Indicator_6.png)

**Reply by fabianmuehlberger** [reply]
I also think that the communication between the two IC's would be the most challenging part, despite the display itself. 

The fantastic thing about [ESPHome](https://esphome.io/) is, that you flash the firmware via [ESPHome CLI ](https://esphome.io/guides/cli.html) or the [esphome web tool](https://web.esphome.io/). After that, the device can create a simple web server to show all relevant information (sensors, switches, logs). 
The configuration itself is then done via one .yaml file that is compiled to a firmware before uploading it to the device. It automagically includes all necessary components that are described in the .yaml and compiles them into one binary that can be uploaded via OTA or USB serial. 

> As for integrating with ESPHome (which I'm not overly familiar with but will try out, as I happen to have a XIAOS3 – haha),

Getting started with the XIAO is fairly easy, and I highly recommend trying it! 

> I'm quite intrigued by your approach, especially the use of pre-compiled firmware and configuration files.

As mentioned above, the ESPHome framework consists of components that are included in the project during compile, custom components or .cpp files are possible too, as well as using lambdas, that allow to write c code inside the .yaml config file. 

- Sensors: The ESPHome framework already has a robust UART communication interface and I assume that the data transfer from the sensors would be a minor problem to solve depending on the data structure. 

- Display: ESPHome is using LVGL too, but I am not familiar with the implementation. I will try to get in contact with the developers (they are a friendly community, :) and see what they think about this Idea. 

I am really excited about the idea of using the indicator with ESPhome!

Best

**Reply by Love4yzp** [reply]
Fantastic idea! Could you please create a repository and share it with me? I'll need some time to get into it. Thank you for sharing!

**Comment by fabianmuehlberger** [comment]
Hey @Love4yzp I asked the ESPHome community on their [Discord](https://discord.com/invite/KhAMKrd), and fortunately, there is already work being done to support the display.  Clydebarrow is working on it on branch [st7701s](https://github.com/clydebarrow/esphome/tree/st7701s). I invited him to join our discussion over here, and I will try to assist where I can. 

Some things I think are still open. 
- It would certainly be helpful to investigate the ESP32s3 to RP2040 communication for providing the sensor data. 
- What is the concept of loading the display .json file from the SD card? Having the esp32 be responsible would make it easier to load the configuration in the same place as the esphome config. Are there any caveats regarding this?

**Comment by Love4yzp** [comment]
Inspired by @Yoobe's project, the ESP32 request the display config file from the RP2040 upon power-up, and then reload the view, seems solid.

**Comment by clydebarrow** [comment]
I have implemented drivers for the Indicator display and touchscreen in ESPHome. Sample config file is here:

https://gist.github.com/clydebarrow/ef89e9a93bd44771483b9144ae9042a1

![indicator (1)](https://github.com/Seeed-Solution/SenseCAP_Indicator_ESP32/assets/2366188/2dea32d1-26d6-4703-837b-54cf77b49f89)

**Comment by jantman** [comment]
It appears that in March 2024 [Seeed announced that the SenseCAP Indicator natively works with ESPHome](https://www.seeedstudio.com/blog/2024/03/28/exciting-news-seeed-studios-mmwave-sensors-and-sensecap-indicatorsupport-esphome/?srsltid=AfmBOoosMlJzZQT95S3e5H6nv38RI1BJtRSd4KnwONakGb_KT0sIsxeF). However, I've been unable to find an example ESPHome configuration file for the SenseCAP Indicator anywhere that includes all of the functionality and uses the latest relevant ESPHome components.

I was wondering if anyone has a fully working, current example config for using the SenseCAP Indicator with ESPHome? i.e. both as a touchscreen display, and as a source for the sensor data?

**Reply by clydebarrow** [reply]
SenseCAP is an umbrella term covering multiple sensors and gateways so there is no such thing as "the SenseCAP". If you are just referring to the _Indicator_, the sensor components in the Indicator that are interfaced via the RP2040 are not currently supported by ESPHome (they probably would be if directly connected to the ESP32.) The display and touchscreen are supported.

**Reply by jantman** [reply]
Apologies for the misunderstanding; yes, I mean the indicator and have edited my post accordingly.

Well... that's a bit of a disappointment, since I just ordered four of them based on the Seeed announcement that they work natively with ESPHome.

Is anyone known to be working on getting at the sensors behind the RP2040?

**Reply by Yoobe** [reply]
It is probably possible to write a component for ESPHome to read the data from the RP2040 (like in the native code for the indicator) and make those sensors available to ESPHome interface. Similar to a driver to read data from i2c device etc.

**Reply by tobi1449** [reply]
> It is probably possible to write a component for ESPHome to read the data from the RP2040 (like in the native code for the indicator) and make those sensors available to ESPHome interface. Similar to a driver to read data from i2c device etc.

I've got a working prototype of an external component that allows reading the sensor data from the RP2040. A few things (like sending the shutdown command) are still missing, but I think I'll be able to publish it in the next few days.

**Reply by tobi1449** [reply]
Here, give it a try: [tobi1449/esphome-sensecap-indicator-sensors](https://github.com/tobi1449/esphome-sensecap-indicator-sensors)

**Reply by jantman** [reply]
@tobi1449 This looks great, thanks so much! Do you happen to have an example esphome config for this, or at least the UART portion that's not in the README of that repo but appears to be required?

**Reply by tobi1449** [reply]
Ah, seems I forgot to include that part, sorry about that :(
I've added it to the usage sample in the README, and I've also included a complete sample config, based on the one @clydebarrow posted above (thanks for that btw!)

**Reply by jantman** [reply]
Thanks so much!!

**Reply by JS876** [reply]
@tobi1449 are you able read other peripherals devices connected to the grove ports using the external component? Example a RFID reader. 

Also, thanks for the prototype, got it working on my D1S.

**Reply by tobi1449** [reply]
> @tobi1449 are you able read other peripherals devices connected to the grove ports using the external component? Example a RFID reader.
> 
> Also, thanks for the prototype, got it working on my D1S.

Not as it is implemented in my repo. It's meant to read from the external sensor included with some of the Indicator models and from the internal SCD41 sensor of the Indicator.
However you might use my [extension of the original firmware file](https://github.com/tobi1449/esphome-sensecap-indicator-sensors/) as inspiration on how to add support for other devices to the internal RP2040 and then modify my esphome component from above to read those as sensors.

Specifically, [see this commit](https://github.com/tobi1449/sensecap-indicator-rp2040-extended/commit/3a6572c690410dea5679c53e9814b47666ae8e8a) where I've added code to send temperature and humidity data from the SCD41 from the RP2040 and [this method](https://github.com/tobi1449/esphome-sensecap-indicator-sensors/blob/1ba408f4114907e8b1ee668a1254e7f22230f391/components/sensecap_indicator_sensors/sensecap_indicator_sensors.cpp#L83) where the incoming packets are read on the ESP32 and the esphome sensors are updated.
